🏁[Cross-check] REST Client

Starkoans opened the pull request:
## 📝 In short
- [Task URL](https://github.com/rolling-scopes-school/tasks/blob/master/react/modules/tasks/final.md)
- [YouTube Video](https://youtu.be/f0qcXN5c3xA) and another one https://www.youtube.com/watch?v=hA8moyuEzYo
- Deploy Link: https://rest-client-app-lac.vercel.app/ru
- Submittion Date: 21.09.2025 / Deadline Date 22.09.2025

## 📸 Screenshots (if UI)
<img width="724" height="74" alt="image" src="https://github.com/user-attachments/assets/bd6ccc3f-b150-4096-b9c0-d216535a8d70" />

## 🧮 Evaluation criteria
Total score: 500 / 500

### Main route - max 50 points

 - [x] The Main page should contain general information about the developers, project, and course. - 10 points
 - [x] In the upper right corner there are 2 buttons: Sign In and Sign Up. - 10 points
 - [x] If the user is authorized, there should be a Main Page button instead of Sign In and Sign Up buttons in the upper right corner. - 10 points
 - [x] If the token is expired/invalid - the user should be redirected from the private routes to the Main page (it might happen either automatically, on page refresh or on route change). - 10 points
 - [x]  Pressing the Sign In / Sign up button redirects a user to the route with the Sign In / Sign up form. - 10 points

### Sign In / Sign Up - max 50 points
 - [x]  Buttons for Sign In / Sign Up / Sign Out are everywhere where they should be. - 10 points
 - [x]  Client-side validation is implemented. - 20 points
 - [x]  Upon successful login, the user is redirected to the Main page. - 10 points
 - [x]  If the user is already logged in and tries to reach these routes, they should be redirected to the Main page. - 10 points

### RESTful client - max 150 points
 - [x] Functional editor enabling query editing and prettifying, request body provided in the URL as base64-encoded on request submit. - 35 points
 - [x] Functional read-only response section, with information about HTTP status and the code. - 30 points
 - [x] Method selector, shows all the valid HTTP verbs, value is provided in the URL on request submit. - 15 points
 - [x] Input for the URL, entered value is provided in base64-encoded way on request submit. - 20 points
 - [x] Headers section, value is provided in the URL on request submit. - 20 points
 - [x] Code generation section. - 30 points

### History and analytics route - max 100 points
 - [x] History and analytics is server-side generated and shows informational message with links to the clients when there are no requests in the database. - 20 points
 - [x] User can navigate to the previously executed HTTP request to the RESTful client, HTTP method, URL, body, headers are restored. - 30 points
 - [x] The following analytics are recorded to the database from the application server side and displayed to the user: request duration, response status code, request timestamp, request method, request size, response size, error details, endpoint/URL. - 50 points

### Variables route - max 50 points
 - [x]  Variables show all the added variables, restores them from the local storage on load. - 15 points
 - [x]  User can add new, or delete an existing variable, variables in the local storage are updated on change. - 15 points
 - [x]  Variables are used in the request before the request execution. - 20 points

### General requirements - max 50 points
 - [x]  Multiple (at lest 2) languages support / i18n. - 30 points
 - [x]  Sticky header. - 10 points
 - [x]  Errors are displayed in the user friendly format. - 10 points
 
### Youtube video - max 50 points

### Penalties
#### 0. Framework Choice
  - [ ] Using any framework option other than the mandatory list (React Router 7 (Framework mode), Next.js (App Router), Tanstack Start, Waku) is strictly forbidden and will result in -200 points
#### 1. TypeScript & Code Quality

 - [ ]  @ts-ignore or any usage (search through GitHub repo) -20 points for each
 - [ ]  The presence of code-smells (God-object, chunks of duplicate code), commented code sections -10 points per each
#### 2. Test Coverage

 - [ ]  Statement coverage below 80% (≥70%): -50 points
 - [ ]  Statement coverage below 70% (≥50%): -100 points
 - [ ]  All coverage metrics below 50%: -150 points
 - [ ]  Absence of tests -250 points
#### 3. React Best Practices

 - [ ]  One of the required lazy-loaded routes isn't lazy-loaded -50 per each
#### 4. Console & Error Handling

 - [ ]  The presence of errors and warnings in the console -20 points for each
 - [ ]  The presence in the console of the results of the console.log execution -20 points for each
 - [ ]  HTTP 4xx and 5xx status codes displayed as errors not in the response section -50 points
#### 5. Development Tools

 - [ ]  Absence of a linting tool -150 points
 - [ ]  Absence of a formatting tool -100 points
 - [ ]  Absence of husky git hooks -100 points
#### 6. UI/UX

 - [ ]  Vite/NextJS default favicon -50 points
#### 7. Project Management
 - [ ] Making commits after the deadline -100 points
 - [ ] Pull Request doesn't follow guideline (including checkboxes in Score) [PR example](https://rs.school/docs/en/pull-request-review-process#pull-request-description-must-contain-the-following) -10 points
 - [ ]  ⚠️ The administration reserves the right to apply penalties for the use of incorrect repository or branch names

aissatsana committed:
fix: fix displaying custom 404 page

Commit:
fix: resolve conflicts

Ulistonee committed:
feat: add info

Commit:
feat: variables component

Ulistonee committed:
feat: add photos

aissatsana committed:
feat: add storing request data to db

aissatsana committed:
feat: add fetching user history

Ulistonee committed:
refactor: as per mentor review

Ulistonee committed:
refactor: as per mentor review

Ulistonee committed:
refactor: as per mentor review

Ulistonee committed:
refactor: as per mentor review

Ulistonee committed:
refactor: as per mentor review

aissatsana committed:
feat: add displaying info

aissatsana committed:
feat: add translations

aissatsana committed:
feat: add restoring request by link

Commit:
feat: variables store

Commit:
refactor: add variable

aissatsana committed:
chore: ignore firebase service account json

Commit:
feat:  request transformer with variables

Commit:
refactor: file structure

Commit:
refactor: response store

Commit:
refactor: cleanup

aissatsana committed:
fix: resolve conflicts

aissatsana committed:
fix: corrects according to review

aissatsana committed:
fix: change data for logging

Commit:
refactor: pr comments

Commit:
Merge branch 'develop' into feat/main-page

Commit:
feat: add logo

Commit:
fix: test

Starkoans committed:
feat/main page (#18)

## 📝 Description
<!-- Describe what changed, why -->

- [x] info about developers, project, course

## 📸 Screenshots (if UI)
<img width="1165" height="657" alt="Screenshot 2025-09-11 at 01 18 01" src="https://github.com/user-attachments/assets/ab28c073-1e59-4cf2-ab68-106ffb10b4c9" />


## 🔗 Related tasks
Closes [this](https://trello.com/c/XwBxeSoB/30-the-main-page-should-contain-general-information-about-the-developers-project-and-course)

Commit:
Merge branch 'develop' of https://github.com/Laser-sharks-react/rest-client-app into refactor/for-mentor-review

Commit:
refactor: request store

Starkoans committed:
Refactor/for mentor review (#19)

## 📝 Description
<!-- Describe what changed, why -->
 changes according to mentor's review. Now the request is being sent on server side and is rendered on the page.

## 📸 Screenshots (if UI)
<img width="1040" height="738" alt="Screenshot 2025-09-13 at 22 20 04" src="https://github.com/user-attachments/assets/aa7e4cca-83db-4ef3-8630-8fc863dd6974" />


## 🔗 Related tasks
closes [this](https://trello.com/c/E507nmDS/40-%D0%BF%D1%80%D0%B0%D0%B2%D0%BA%D0%B8-%D0%BE%D1%82-%D0%BB%D0%B5%D1%88%D0%B8)

Commit:
refactor: variables store

Commit:
Merge branch 'develop' of https://github.com/Laser-sharks-react/rest-client-app into feat/variables

Commit:
fix: request store clear headers

Commit:
refactor: layout, fetchProxyRequest

Commit:
feat: dynamic import for variables

Commit:
refactor: login, sign up  forms

Commit:
refactor: history page, request list table

Starkoans committed:
Feat/variables (#20)

## 📝 Description
<!-- Describe what changed, why -->

- [x] This route should be private.
- [x] Variables code should be lazy-loaded (so unauthenticated user
won't download the code).
- [x] Variables should be stored in the browser's local storage.
- [x] Application should load previously saved variables on user's
successful authentication.
- [x] Variable value can be inserted either in the url, or request body,
or header value.
- [x] To indicate that the variable should be used, user can specify it
as {{variableName}}.
- [x] Variable in the url: https://wwww.npmjs.com/{{packageName}}
- [x] Variable in the header: {{APPLICATION_JSON}}
- [x] Variable in the JSON


## 📸 Screenshots (if UI)
<!-- Insert before/after screenshots -->
<img width="866" height="867" alt="image"
src="https://github.com/user-attachments/assets/81d342d8-10ee-4018-b423-c496fcca65fb"
/>

## 🔗 Related tasks
<!-- Specify the issue number, for example: Closes #123 -->
https://trello.com/c/SbyJ05Rv/28-variables

Ulistonee committed:
fix: fix link to request page

Ulistonee committed:
chore: merge with remote branch

Commit:
chore: test ignore paths

Commit:
feat: get variables from ls tests

Commit:
feat: replace variables tests

Ulistonee committed:
refactor: refactor request record type

Ulistonee committed:
Merge branch 'develop' into feat/history-page-rebased

# Conflicts:
#	.env
#	messages/en.json
#	messages/ru.json
#	src/app/api/proxy/route.ts
#	src/components/header.tsx

Ulistonee committed:
Feat/history page (#21)

## 📝 Description
- [x] This route should be private. 
- [x] History and analytics code should be lazy-loaded (so
unauthenticated user won't download the code).
- [x] The section should be server-side generated: all request history
and analytics are aggregated and rendered on the server, then sent to
the client for display.
- [x] The section should display requests using links, on clicking on
the link, user should be navigated to the respective section (RESTful
client).
- [x] All request history and analytics information must be loaded from
Firebase, Supabase, or Convex for the logged in user. The information is
written to the database from the server side after each request.
- [x] The following parameters must be stored for each request: Request
Duration (Latency): Time taken from sending the request to receiving the
response.
Response Status Code: HTTP status code (e.g., 200, 404, 500) for each
request.
Request Timestamp: When the request was made (date and time).
Request Method: GET, POST, PUT, DELETE, etc.
Request Size: Size of the request payload (in bytes).
Response Size: Size of the response payload (in bytes).
Error Details: If the request failed, log the error message or type
(timeout, network error, etc.).
Endpoint/URL: The endpoint that was called (for grouping and analysis).
- [x] The section should show requests sorted by the time of their
execution.
- [x] The section should not contain variables, because information
should be recorded from the server side.
- [x] After navigating to the respective section, all the functional
fields (URL, method selector, headers, body, values) should be restored.
Please, mind this when saving the request in the database to ensure that
all the required data will be saved.
- [x] If there are no requests in the history, show message to the user,
e.g. "You haven't executed any requests yet", "It's empty here. Try
those options:" and give links to the RESTful client.


## 📸 Screenshots (if UI)
<img width="400" alt="image"
src="https://github.com/user-attachments/assets/49126e3d-c2f3-40f5-babe-75bd10048854"
/>


## 🔗 Related tasks
https://trello.com/c/OGdq3JiN/29-history-and-analytics

Ulistonee committed:
fix: body in get requests

Commit:
feat: test for access routes

Commit:
feat: tests for base 64 utils

Commit:
feat: tests for code generator

Commit:
feat: tests for headers to arr, is http method, parse path

aissatsana committed:
test: add tests for stores

aissatsana committed:
test: add tests for main, hot-found, error pages

Ulistonee committed:
fix: fix cookies on server

aissatsana committed:
test: add minimal tests for login and signup pages

Ulistonee committed:
fix: fix small issues

aissatsana committed:
test: add tests for components

Commit:
feat: fetch proxy request tests

Commit:
refactor: remove duplicate util

Commit:
refactor: with utils and constants

Starkoans committed:
Fix/client (#23)

## 📝 Description

- [x] fix issue with authorization error
- [x] fix validation of private and public routes
- [x] fix empty url fetching

## 📸 Screenshots (if UI)
<img width="1122" height="621" alt="Screenshot 2025-09-21 at 01 52 33" src="https://github.com/user-attachments/assets/ee2d25ad-bd7e-4efe-8959-380fdc8d285e" />


## 🔗 Related tasks

Commit:
Merge branch 'develop' of https://github.com/Laser-sharks-react/rest-client-app into feat/tests

Commit:
fix: custom response tests types

aissatsana committed:
test: complete tests for components

aissatsana committed:
test: add test for history page

Commit:
test: session utils

Commit:
Merge branch 'feat/tests' of https://github.com/Laser-sharks-react/rest-client-app into feat/tests

Commit:
Merge branch 'feat/tests' of https://github.com/Laser-sharks-react/rest-client-app into feat/tests

Ulistonee committed:
fix: fix /en.fsdf path for 404 page

Ulistonee committed:
fix: fix /en.fsdf path for 404 page

Commit:
test: test with coverage over 80

Commit:
fix: link mock

Commit:
remove: md file

Commit:
Merge branches 'feat/tests' and 'develop' of https://github.com/Laser-sharks-react/rest-client-app into feat/tests

Ulistonee committed:
fix: fix main page welcome title

Ulistonee committed:
fix: fix main page welcome title

Ulistonee committed:
fix: fix main page welcome title

Ulistonee committed:
fix: fix body removal on get method

Ulistonee committed:
feat: add favicon

Starkoans committed:
Feat/tests (#22)

## 📸 Screenshots (if UI)
<img width="734" height="257" alt="image"
src="https://github.com/user-attachments/assets/04a5f7fe-2897-4988-9de1-71360f4e619f"
/>

## 🔗 Related tasks
<!-- Specify the issue number, for example: Closes #123 -->

Commit:
Merge branch 'develop' into fix/fixes

Commit:
fix: tests

Starkoans committed:
Fix/fixes (#25)

## 📝 Description
- [x] fix 404 page for /en.asdfasdf
- [x] If logged in, there shouldn't be any sign-in or sign-up buttons,
and there should be a Welcome Back, [Username] line
- [x] Bug: When switching between tabs and selecting the GET method, the
body is still sent with the request
- [x] Add a favicon.

## 📸 Screenshots (if UI)
<img width="837" height="504" alt="Screenshot 2025-09-21 at 19 28 27"
src="https://github.com/user-attachments/assets/fe601659-c5c9-48cf-bbcf-12893f8322a1"
/>


## 🔗 Related tasks

Ulistonee committed:
fix: add sign up btn

Starkoans committed:
fix: add sign up btn (#27)

## 📝 Description
<!-- Describe what changed, why -->

- [x] add signup btn to header 


## 📸 Screenshots (if UI)
<img width="1448" height="79" alt="Screenshot 2025-09-21 at 20 24 51"
src="https://github.com/user-attachments/assets/7f3ae9da-bb74-4487-9cfb-d942d1b41a53"
/>


## 🔗 Related tasks

Commit:
fix: prefetch links in navbar

Starkoans committed:
fix: prefetch links in navbar (#29)

## 📝 Description
<!-- Describe what changed, why -->

- [ ] 
- [ ]
- [ ]

## 📸 Screenshots (if UI)
<!-- Insert before/after screenshots -->

## 🔗 Related tasks
<!-- Specify the issue number, for example: Closes #123 -->

aissatsana committed:
feat: add link to home page in upper right corner

aissatsana committed:
feat: fix width for forms

aissatsana committed:
fix: change visibility of errors

a-rudak-EIS commented on src/app/[locale]/login/page.tsx:
```diff
@@ -3,12 +3,14 @@
   Button,
+  Card,
   Container,
   FormControl,
   FormHelperText,
   FormLabel,
```
```suggestion
```

a-rudak-EIS commented on src/app/[locale]/login/page.tsx:
```diff
@@ -25,6 +27,7 @@ import { ROUTES } from '@/lib/constants/routes';
 export default function Login() {
   const t = useTranslations('LoginPage');
   const tForm = useTranslations('Form');
+  const tErrors = useTranslations('FormErrors');
```
```suggestion
```

a-rudak-EIS commented on src/app/[locale]/signup/page.tsx:
```diff
@@ -2,12 +2,14 @@
   Button,
+  Card,
   Container,
   FormControl,
   FormHelperText,
   FormLabel,
```
```suggestion
```

a-rudak-EIS commented on src/components/json-editor.tsx:
```diff
@@ -22,13 +23,16 @@ type Props = {
 
 export default function JsonEditor({
-  initial = '{\n  "hello": "world"\n}',
+  initial,
```
было лучше, с 


```suggestion
  initial = '{\n  "hello": "world"\n}',
```

только '{\n  "hello": "world"\n}' бы еще в константу вытащить

Starkoans committed:
Feat/link-to-main (#30)

## 📝 Description
- [x] add link to main page in upper right corner
- [x] fix width for form in login and signup pages

## 📸 Screenshots (if UI)
<img width="400" alt="image"
src="https://github.com/user-attachments/assets/5082e254-a0c0-4317-bdb2-59f81ca6fdfa"
/>
<img width="400" alt="image"
src="https://github.com/user-attachments/assets/a0654686-f014-4ca6-bd1b-dd517191434f"
/>


## 🔗 Related tasks
